### DISCUSSION TITLE
is it possible to compute overlap between CC and CI wavefunctions?

### DISCUSSION THREAD

**Discussion by mkrompiec** [pyscf/pyscf#2630] [Q&A]
There is a function for computing overlap between CI wavefunctions. Is something similar possible with CC? Is there a function for converting CC amplitudes into a CI vector?

**Comment by MatthewRHermes** [comment]
You can probably abuse the `direct_nosym` FCI solver module, which is designed for non-Hermitian Hamiltonians, to transform a (spin-restricted) CC wave function into a full CI vector:

```
import numpy as np
from scipy import linalg
from pyscf import lib
from pyscf.fci import cistring
from pyscf.fci.direct_nosym import absorb_h1e, contract_2e

nmo = nocc + nvir
neleca = nelecb = nocc
ndeta = cistring.num_strings (norb, neleca)
ndetb = cistring.num_strings (norb, nelecb)
fcivec = np.zeros ((ndeta, ndetb), dtype=float)
fcivec[0,0] = 1.0
link_indexa = cistring.gen_linkstr_index(range(norb), neleca)
link_indexb = cistring.gen_linkstr_index(range(norb), nelecb)
linkstr = (link_indexa, link_indexb)

t1 = np.zeros ((nmo, nmo), dtype=mycc.t1.dtype)
t1[nocc:,:nocc] = mycc.t1[:,:].T
t2 = np.zeros ((nmo, nmo, nmo, nmo), dtype=mycc.t2.dtype)
t2[nocc:,:nocc,nocc:,:nocc] = mycc.t2.transpose (2,0,3,1)
T = absorb_h1e (t1, t2, norb, (neleca,nelecb), fac=0.5)
# I think absorb_h1e will multiply t1 by 2 by default. 
# fac=0.5 cancels that, but it also multiplies t2 by 0.5
# So maybe you have to multiply t2 by 2?
# Or not, I'm not sure how factors of 1/2 work in the cc module

x = fcivec
nfac = 1
for n in range (nocc+1):
    x = contract_2e (T, x, norb, (neleca,nelecb), link_index=linkstr)
    nfac *= n
    fcivec += x / nfac

epsilon = 1e-8
# If this works the way I think, x should be a zero vector now
assert (linalg.norm (x) < epsilon)
# If this works the way I think, fcivec should still be in intermediate normalization
assert (abs (fcivec[0,0] - 1.0) < epsilon)
# If this works the way I think, |<hf|ci'ca|fcivec>| = |t1[i,a]|
t1_ref = np.abs (mycc.t1)
strs_t1a = cistring.tn_strs (norb, neleca, 1)
addr_t1a = cistring.strs2addr (norb, neleca, strs_t1a)
t1a_test = np.abs (fcivec[addr_t1a,0].reshape (nocc,nvir))
assert (np.amax (np.abs (t1a_test - t1_ref)) < epsilon)
strs_t1b = cistring.tn_strs (norb, nelecb, 1)
addr_t1b = cistring.strs2addr (norb, nelecb, strs_t1b)
t1b_test = np.abs (fcivec[0,addr_t1b].reshape (nocc,nvir))
assert (np.amax (np.abs (t1b_test - t1_ref)) < epsilon)
# There is another test you can make for the signs, and for t2, but it's more complicated
```

**Comment by MatthewRHermes** [comment]
The existence of the function `cistring.tn_strs` suggests that someone was working on this functionality at some point, but it doesn't seem to be available in the box yet.

**Comment by obackhouse** [comment]
I have conversion routines between T amplitudes, C amplitudes, and CI vectors implemented up to fourth order in my package `ebcc`, which extends PySCF: https://github.com/BoothGroup/ebcc/blob/b7df8ae0d147844539146e8e68965bfa91b451fd/ebcc/ext/fci.py#L199

I would recommend the unrestricted version for higher than second order. Any questions are probably best as new issues on my repo.

**Comment by Physfock** [comment]
> I have conversion routines between T amplitudes, C amplitudes, and CI vectors implemented up to fourth order in my package `ebcc`, which extends PySCF: https://github.com/BoothGroup/ebcc/blob/b7df8ae0d147844539146e8e68965bfa91b451fd/ebcc/ext/fci.py#L199
> 
> I would recommend the unrestricted version for higher than second order. Any questions are probably best as new issues on my repo.

Can you provide an example of hot to use this to calculate the overlap between CC wave functions?

**Comment by obackhouse** [comment]
> > I have conversion routines between T amplitudes, C amplitudes, and CI vectors implemented up to fourth order in my package `ebcc`, which extends PySCF: https://github.com/BoothGroup/ebcc/blob/b7df8ae0d147844539146e8e68965bfa91b451fd/ebcc/ext/fci.py#L199
> > I would recommend the unrestricted version for higher than second order. Any questions are probably best as new issues on my repo.
> 
> Can you provide an example of hot to use this to calculate the overlap between CC wave functions?

Do you mean starting with a pair of PySCF RCCSD classes? are they for the same system/reference state? it might be 
 better if you open an issue on my repo so as not to go off-topic on the PySCF repo, if you don't mind.
